OutOfMemoryError when de-serializing the input for SynthesizeDependenciesTask

**Build scan link**
Private, can't share unfortunately.

**Plugin version**
`v1.17.0`

**Gradle version**
`7.4.2`

**(Optional) Android Gradle Plugin (AGP) version**
Not applicable.

**`reason` output for bugs relating to incorrect advice**
Not applicable.

**Describe the bug**
An `OutOfMemoryError` is sometimes thrown when de-serializing the input for `SynthesizeDependenciesTask`
```
Caused by: java.lang.OutOfMemoryError: Java heap space
at kotlin.io.TextStreamsKt.copyTo(ReadWrite.kt:125)
at kotlin.io.TextStreamsKt.copyTo$default(ReadWrite.kt:120)
at kotlin.io.TextStreamsKt.readText(ReadWrite.kt:107)
at kotlin.io.FilesKt__FileReadWriteKt.readText(FileReadWrite.kt:125)
at kotlin.io.FilesKt__FileReadWriteKt.readText$default(FileReadWrite.kt:125)
at com.autonomousapps.tasks.SynthesizeDependenciesTask$SynthesizeDependenciesWorkAction.execute(SynthesizeDependenciesTask.kt:245)
```

**To Reproduce**
It's not happening deterministically unfortunately, but having an insane number of dependencies sometimes triggers this.

**Expected behavior**
No OOM exception.

**Additional context**
Similar to #823 where it was happening when serializing, the de-serializing is done without buffering.

--- a/src/main/kotlin/com/autonomousapps/internal/utils/utils.kt
+++ b/src/main/kotlin/com/autonomousapps/internal/utils/utils.kt
@@ -2,6 +2,8 @@
 
 package com.autonomousapps.internal.utils
 
+import okio.buffer
+import okio.source
 import org.gradle.api.artifacts.result.DependencyResult
 import org.gradle.api.artifacts.result.ResolvedDependencyResult
 import org.gradle.api.file.RegularFile
@@ -20,6 +22,20 @@ internal fun RegularFileProperty.getAndDelete(): File {
 
 internal inline fun <reified T> RegularFileProperty.fromJsonSet(): Set<T> = get().fromJsonSet()
 internal inline fun <reified T> RegularFile.fromJsonSet(): Set<T> = asFile.readText().fromJsonSet()
+
+/**
+ * Buffers reads of the RegularFileProperty from disk to the set.
+ */
+internal inline fun <reified T> RegularFileProperty.bufferReadJsonSet(): Set<T> = get().bufferReadJsonSet()
+
+/**
+ * Buffers reads of the RegularFile from disk to the set.
+ */
+internal inline fun <reified T> RegularFile.bufferReadJsonSet(): Set<T> {
+  asFile.source().buffer().use { reader ->
+    return getJsonSetAdapter<T>().fromJson(reader)!!
+  }
+}
 
 internal inline fun <reified T> RegularFileProperty.fromJsonList(): List<T> = get().fromJsonList()
 internal inline fun <reified T> RegularFile.fromJsonList(): List<T> = asFile.readText().fromJsonList()
@@ -58,6 +74,27 @@ internal inline fun <reified T> File.fromJson(): T = readText().fromJson()
 
 internal inline fun <reified T> RegularFileProperty.fromNullableJsonSet(): Set<T>? {
   return orNull?.asFile?.readText()?.fromJsonSet()
+}
+
+/**
+ * Buffer reads of the RegularFileProperty from disk to the set.
+ */
+internal inline fun <reified T> RegularFileProperty.bufferReadJson(): T = get().bufferReadJson()
+
+/**
+ * Buffer reads of the RegularFile from disk to the set.
+ */
+internal inline fun <reified T> RegularFile.bufferReadJson(): T {
+  asFile.source().buffer().use { reader ->
+    return getJsonAdapter<T>().fromJson(reader)!!
+  }
+}
+
+/**
+ * Buffer reads of the nullable RegularFileProperty from disk to the set.
+ */
+internal inline fun <reified T> RegularFileProperty.bufferReadNullableJsonSet(): Set<T> {
+  return orNull?.bufferReadJsonSet() ?: emptySet()
 }
 
 internal fun RegularFileProperty.readLines(): List<String> = get().readLines()

--- a/src/main/kotlin/com/autonomousapps/tasks/SynthesizeDependenciesTask.kt
+++ b/src/main/kotlin/com/autonomousapps/tasks/SynthesizeDependenciesTask.kt
@@ -126,17 +126,17 @@ abstract class SynthesizeDependenciesTask @Inject constructor(
     override fun execute() {
       val outputDir = parameters.outputDir
 
-      val dependencies = parameters.compileDependencies.fromJson<CoordinatesContainer>().coordinates
-      val physicalArtifacts = parameters.physicalArtifacts.fromJsonSet<PhysicalArtifact>()
-      val explodedJars = parameters.explodedJars.fromJsonSet<ExplodedJar>()
-      val inlineMembers = parameters.inlineMembers.fromJsonSet<InlineMemberDependency>()
-      val serviceLoaders = parameters.serviceLoaders.fromJsonSet<ServiceLoaderDependency>()
-      val annotationProcessors = parameters.annotationProcessors.fromJsonSet<AnnotationProcessorDependency>()
+      val dependencies = parameters.compileDependencies.bufferReadJson<CoordinatesContainer>().coordinates
+      val physicalArtifacts = parameters.physicalArtifacts.bufferReadJsonSet<PhysicalArtifact>()
+      val explodedJars = parameters.explodedJars.bufferReadJsonSet<ExplodedJar>()
+      val inlineMembers = parameters.inlineMembers.bufferReadJsonSet<InlineMemberDependency>()
+      val serviceLoaders = parameters.serviceLoaders.bufferReadJsonSet<ServiceLoaderDependency>()
+      val annotationProcessors = parameters.annotationProcessors.bufferReadJsonSet<AnnotationProcessorDependency>()
       // Android-specific and therefore optional
-      val manifestComponents = parameters.manifestComponents.fromNullableJsonSet<AndroidManifestDependency>().orEmpty()
-      val androidRes = parameters.androidRes.fromNullableJsonSet<AndroidResDependency>().orEmpty()
-      val nativeLibs = parameters.nativeLibs.fromNullableJsonSet<NativeLibDependency>().orEmpty()
-      val androidAssets = parameters.androidAssets.fromNullableJsonSet<AndroidAssetDependency>().orEmpty()
+      val manifestComponents = parameters.manifestComponents.bufferReadNullableJsonSet<AndroidManifestDependency>()
+      val androidRes = parameters.androidRes.bufferReadNullableJsonSet<AndroidResDependency>()
+      val nativeLibs = parameters.nativeLibs.bufferReadNullableJsonSet<NativeLibDependency>()
+      val androidAssets = parameters.androidAssets.bufferReadNullableJsonSet<AndroidAssetDependency>()
 
       physicalArtifacts.forEach { artifact ->
         builders.merge(

